### ISSUE TITLE
Clang crashes when using asm goto with matching constraints

### ISSUE THREAD

**Issue by josephcsible** [llvm/llvm-project#53547]
Minimal reproducer:
```c
int main(void) {
    int in = 0;
    int out;
    asm goto("" : "=r" (out) : "0" (in) : : label);
    return out;
    label:
    return 0;
}
```
Result:
```
fatal error: error in backend: Cannot emit physreg copy instruction
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: /opt/compiler-explorer/clang-13.0.0/bin/clang -g -o /app/output.s -S --gcc-toolchain=/opt/compiler-explorer/gcc-11.2.0 -fcolor-diagnostics -fno-crash-diagnostics <source>
1.	<eof> parser at end of file
2.	Code generation
3.	Running pass 'Function Pass Manager' on module '<source>'.
4.	Running pass 'Post-RA pseudo instruction expansion pass' on function '@main'
 #0 0x0000564b11cdf7ff PrintStackTraceSignalHandler(void*) Signals.cpp:0:0
 #1 0x0000564b11cdd6c0 llvm::sys::CleanupOnSignal(unsigned long) (/opt/compiler-explorer/clang-13.0.0/bin/clang+0x34dd6c0)
 #2 0x0000564b11c2ee12 llvm::CrashRecoveryContext::HandleExit(int) (/opt/compiler-explorer/clang-13.0.0/bin/clang+0x342ee12)
 #3 0x0000564b11cd628e llvm::sys::Process::Exit(int, bool) (/opt/compiler-explorer/clang-13.0.0/bin/clang+0x34d628e)
 #4 0x0000564b0fa8469e LLVMErrorHandler(void*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool) cc1_main.cpp:0:0
 #5 0x0000564b11c36299 llvm::report_fatal_error(llvm::Twine const&, bool) (/opt/compiler-explorer/clang-13.0.0/bin/clang+0x3436299)
 #6 0x0000564b11c36458 (/opt/compiler-explorer/clang-13.0.0/bin/clang+0x3436458)
 #7 0x0000564b10b2a671 llvm::X86InstrInfo::copyPhysReg(llvm::MachineBasicBlock&, llvm::MachineInstrBundleIterator<llvm::MachineInstr, false>, llvm::DebugLoc const&, llvm::MCRegister, llvm::MCRegister, bool) const (/opt/compiler-explorer/clang-13.0.0/bin/clang+0x232a671)
 #8 0x0000564b1131b23d (anonymous namespace)::ExpandPostRA::runOnMachineFunction(llvm::MachineFunction&) ExpandPostRAPseudos.cpp:0:0
 #9 0x0000564b11072918 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (/opt/compiler-explorer/clang-13.0.0/bin/clang+0x2872918)
#10 0x0000564b114a6eb9 llvm::FPPassManager::runOnFunction(llvm::Function&) (/opt/compiler-explorer/clang-13.0.0/bin/clang+0x2ca6eb9)
#11 0x0000564b114a7151 llvm::FPPassManager::runOnModule(llvm::Module&) (/opt/compiler-explorer/clang-13.0.0/bin/clang+0x2ca7151)
#12 0x0000564b114a83c7 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/opt/compiler-explorer/clang-13.0.0/bin/clang+0x2ca83c7)
#13 0x0000564b11fd29f2 (anonymous namespace)::EmitAssemblyHelper::EmitAssemblyWithNewPassManager(clang::BackendAction, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream> >) BackendUtil.cpp:0:0
#14 0x0000564b11fd6ef5 clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream> >) (/opt/compiler-explorer/clang-13.0.0/bin/clang+0x37d6ef5)
#15 0x0000564b12c33e1d clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) (/opt/compiler-explorer/clang-13.0.0/bin/clang+0x4433e1d)
#16 0x0000564b13a8cd49 clang::ParseAST(clang::Sema&, bool, bool) (/opt/compiler-explorer/clang-13.0.0/bin/clang+0x528cd49)
#17 0x0000564b12c33ec2 clang::CodeGenAction::ExecuteAction() (/opt/compiler-explorer/clang-13.0.0/bin/clang+0x4433ec2)
#18 0x0000564b125e1091 clang::FrontendAction::Execute() (/opt/compiler-explorer/clang-13.0.0/bin/clang+0x3de1091)
#19 0x0000564b1257f352 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/opt/compiler-explorer/clang-13.0.0/bin/clang+0x3d7f352)
#20 0x0000564b126ad7d3 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/opt/compiler-explorer/clang-13.0.0/bin/clang+0x3ead7d3)
#21 0x0000564b0fa8601c cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/opt/compiler-explorer/clang-13.0.0/bin/clang+0x128601c)
#22 0x0000564b0fa821fd ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&) driver.cpp:0:0
#23 0x0000564b1242aaa5 void llvm::function_ref<void ()>::callback_fn<clang::driver::CC1Command::Execute(llvm::ArrayRef<llvm::Optional<llvm::StringRef> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, bool*) const::'lambda'()>(long) Job.cpp:0:0
#24 0x0000564b11c2eca3 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/opt/compiler-explorer/clang-13.0.0/bin/clang+0x342eca3)
#25 0x0000564b1242cb5e clang::driver::CC1Command::Execute(llvm::ArrayRef<llvm::Optional<llvm::StringRef> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, bool*) const (/opt/compiler-explorer/clang-13.0.0/bin/clang+0x3c2cb5e)
#26 0x0000564b1240258a clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&) const (/opt/compiler-explorer/clang-13.0.0/bin/clang+0x3c0258a)
#27 0x0000564b124030df clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*> >&) const (/opt/compiler-explorer/clang-13.0.0/bin/clang+0x3c030df)
#28 0x0000564b1240c4d5 clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*> >&) (/opt/compiler-explorer/clang-13.0.0/bin/clang+0x3c0c4d5)
#29 0x0000564b0f99b46e main (/opt/compiler-explorer/clang-13.0.0/bin/clang+0x119b46e)
#30 0x00007fafb8c800b3 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b3)
#31 0x0000564b0fa81d7a _start (/opt/compiler-explorer/clang-13.0.0/bin/clang+0x1281d7a)
clang-13: error: clang frontend command failed with exit code 70 (use -v to see invocation)
```
https://godbolt.org/z/74fM37434

**Comment by efriedma-quic** [comment]
@nickdesaulniers

**Comment by nickdesaulniers** [comment]
Hi @josephcsible , thanks for the report. I noticed you're using `clang-13`.

> 0.	Program arguments: /opt/compiler-explorer/clang-13.0.0

but it's no longer reproducible on trunk.  This [godbolt link](https://godbolt.org/z/oo5j5rhhx) demonstrates that your reduced test case crashes clang-13, but works in trunk (note: the release/14.x branch was just cut from trunk; clang-14 doesn't exist in release form yet, and trunk technically will become clang-15 at the time of this writing).

I believe [I fixed this recently](https://github.com/llvm/llvm-project/commit/4edb9983cb8c8b850083ee5941468f5d0ef6fe2c) in clang-14.

```c
int main(void) {
    int out = 0;
    asm goto("" : "+r" (out) : : : label);
    return out;
    label:
    return 0;
}
```
might work for you in clang-13, otherwise clang-14 should have a fix. (I can run a full bisection if necessary to pinpoint the fix, but the fix I linked to above is in the release/14.x branch and will simply ride the trains to release).
